<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query_3_1 = f'''select
                count(id)
                from vacancies
'''

In [5]:
df = pd.read_sql_query(query_3_1, connection)
df

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''select
                count(id)
                from employers
'''

In [7]:
df_1 = pd.read_sql_query(query_3_2, connection)
df_1

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/4246428027.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_1 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''select
                count(id)
                from areas
'''

In [9]:
df_2 = pd.read_sql_query(query_3_3, connection)
df_2

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/3399235352.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_2 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''select
                count(id)
                from industries
'''

In [11]:
df_3 = pd.read_sql_query(query_3_4, connection)
df_3

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1492296125.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

**Выводы по предварительному анализу данных:** база данных содержит 47187 вакансий, в которых представлены 23501 работодатель по 294 сферам деястельности из 1362 регионов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''select
                a.name area,
                count(v.id) cnt
                from areas a
                    join vacancies v on a.id = v.area_id
                group by 1
                order by 2 desc'''

In [13]:
df_4 = pd.read_sql_query(query_4_1, connection)
df_4

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/392072453.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4 = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''select
                count(*)
                from vacancies
                where salary_from is not null or salary_to is not null'''

In [15]:
df_5 = pd.read_sql_query(query_4_2, connection)
df_5

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1568068158.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query_4_3 = f'''select
                round(avg(salary_from), 0) as from,
                round(avg(salary_to), 0) as to
                from vacancies'''

In [17]:
df_6 = pd.read_sql_query(query_4_3, connection)
df_6

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2188318473.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6 = pd.read_sql_query(query_4_3, connection)


from        to
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
query_4_4 = f'''select
                schedule,
                employment,
                count(id)
                from vacancies
                group by 1, 2
                order by 3 desc'''

In [19]:
df_7 = pd.read_sql_query(query_4_4, connection)
df_7

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/640631597.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query_4_5 = f'''select
                experience,
                count(id)
                from vacancies
                group by 1
                order by 2'''

In [21]:
df_8 = pd.read_sql_query(query_4_5, connection)
df_8

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/4093658911.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_8 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Выводы по детальному анализу данных:** большая часть вакансий относится к городам-миллионникам. Примерно в половине вакансий обозначена зарплата для соискателей. Примерный разброс величины зарплаты составялет от 70 до 110 т.руб. Наибольшее число вакансий с традиционным рабочим графиком, а именно полный рабочий день на рабочем месте, также заметное число вакансий с возможностью удетеленной работы. Около половины вакансий для соискателей с мниимальным опытом работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
query_5_1 = f'''select
            e.name,
            count(v.id)
            from employers e
            join vacancies v on e.id = v.employer_id
            group by 1
            order by 2 desc
            limit 5'''

In [23]:
df_9 = pd.read_sql_query(query_5_1, connection)
df_9

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1022043397.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_9 = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
query_5_2 = f'''select
            a.name,
            count(e.id) as cnt_employers,
            count(v.id) as cnt_vacancies
            from areas a
            left join employers e on a.id = e.area
            left join vacancies v on v.area_id = a.id
            where v.id is null
            group by a.id
            order by 2 desc'''

In [25]:
df_10 = pd.read_sql_query(query_5_2, connection)
df_10

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/131413738.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_10 = pd.read_sql_query(query_5_2, connection)


name  cnt_employers  cnt_vacancies
0                Россия            410              0
1             Казахстан            207              0
2    Московская область             75              0
3    Краснодарский край             19              0
4    Ростовская область             18              0
..                  ...            ...            ...
588             Скидель              0              0
589          Белоозерск              0              0
590           Орловский              0              0
591          Павловская              0              0
592             Таксимо              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
query_5_3 = f'''select
            e.name,
            count(distinct(v.area_id))
            from employers e
            left join vacancies v on e.id = v.employer_id
            left join areas a on v.area_id = a.id
            group by 1
            order by 2 desc'''


In [27]:
df_11 = pd.read_sql_query(query_5_3, connection)
df_11

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1866768529.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_11 = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23170                  СДЕЛКА      0
23171            Alandr Group      0
23172                     СДК      0
23173      Lemon Land Lombard      0
23174                  Махина      0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
query_5_4 = f'''select
            count(e.id)
            from employers e
            left join employers_industries ei on e.id = ei.employer_id
            where ei.industry_id is null
            '''


In [29]:
df_12 = pd.read_sql_query(query_5_4, connection)
df_12

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1302209370.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_12 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
query_5_5 = f'''select
            ei.employer_id,
            e.name,
            count(ei.industry_id) cnt_industies
            from employers_industries ei
            left join employers e on ei.employer_id = e.id
            group by 1, 2
            having count(ei.industry_id) = 4
            order by 2
            limit 1
            offset 2
            '''

In [31]:
df_13 = pd.read_sql_query(query_5_5, connection)
df_13

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2481876866.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_13 = pd.read_sql_query(query_5_5, connection)


employer_id  name  cnt_industies
0        64174  2ГИС              4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
query_5_6 = f'''select
            count(e.id)
            from employers e
            left join employers_industries ei on e.id = ei.employer_id
            left join industries i on ei.industry_id = i.id
            where i.name = 'Разработка программного обеспечения'
            '''

In [33]:
df_14 = pd.read_sql_query(query_5_6, connection)
df_14

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1298417458.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_14 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
import requests 
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
cities = page.find('table', class_= 'standard').find('tbody').find_all('tr')
cities_list = []
for i in cities:
    cities_list.append(str(i).split('title="')[1].split('">')[0])
cities_list.pop(0)
cities_list = tuple(cities_list)

print(cities_list)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [35]:
query_5_7 = f'''
            select
            a.name cities,
            count(v.id) quantity
            from employers e
            left join vacancies v on e.id = v.employer_id
            left join areas a on v.area_id = a.id
            where e.name = 'Яндекс' and a.name in {cities_list}
            group by a.id
            union
            select
            'total',
            count(v.id)
            from employers e
            left join vacancies v on e.id = v.employer_id
            left join areas a on v.area_id = a.id
            where e.name = 'Яндекс' and a.name in {cities_list}
            group by e.id
            order by 2
            '''

In [36]:
df_15 = pd.read_sql_query(query_5_7, connection)
df_15

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/43371686.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_15 = pd.read_sql_query(query_5_7, connection)


cities  quantity
0              Омск        21
1         Челябинск        22
2        Красноярск        23
3         Волгоград        24
4             Пермь        25
5            Казань        25
6    Ростов-на-Дону        25
7               Уфа        26
8            Самара        26
9         Краснодар        30
10          Воронеж        32
11      Новосибирск        35
12  Нижний Новгород        36
13     Екатеринбург        39
14  Санкт-Петербург        42
15           Москва        54
16            total       485

***

**Выводы по анализу работодателей:**
1. Наибольшее число вакансий у компании Яндекс, на втором месте компения Ростелеком. 
2. Среди регионов, в которых наибольшее число работодателей без вакансий лидирует Россия.
3. Компании Яндекс требуются сотрудники в наибольшем числе регионов, на втором месте компения Ростелеком.
4. У 8419 Работодателей не указана сфера деятельности.
5. Компании 2ГИС требуется специалисты в 4х сферах деятельности.
6. У 3553 компаний указана сфера деятельности как разработка ПО.
7. Компании Яндекс требуются специалисты во всех городах-милионниках.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
query_6_1 = f'''select
            count(name)
            from vacancies
            where lower(name) like '%data%' or lower(name) like '%данн%'
            '''

In [38]:
df_16 = pd.read_sql_query(query_6_1, connection)
df_16

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2378802155.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_16 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
query_6_2 = f'''select
            count(id)
            from vacancies
            WHERE ((name ILIKE '%data scientist%') 
            OR (name ILIKE '%data science%') 
            OR (name ILIKE '%исследователь данных%')
            OR ((name LIKE '%ML%') AND (name NOT LIKE '%HTML%'))
            OR (name ILIKE '%machine learning%')
            OR (name ILIKE '%машинн%обучен%'))
            AND (name ILIKE '%junior%'
            OR experience ILIKE '%Нет опыта%'
            OR employment ILIKE '%Стажировка%')
            '''

In [40]:
df_17 = pd.read_sql_query(query_6_2, connection)
df_17

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2271740752.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_17 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
query_6_3 = f'''SELECT count(id)
                FROM vacancies
                WHERE (name ILIKE '%data scientist%' 
                or name ILIKE '%data science%' 
                or name ILIKE '%исследователь данных%' 
                or name ILIKE '%ml%' and name NOT LIKE '%HTML%' and name not like '%QML%' 
                or name ILIKE '%machine learning%' 
                or name ILIKE '%машинн%обучен%') 
                and(key_skills ILIKE '%postgres%'
                or key_skills ILIKE '%sql%')
            '''

In [42]:
df_18 = pd.read_sql_query(query_6_3, connection)
df_18

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2497234423.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_18 = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
query_6_4 = f'''SELECT count(id)
                FROM vacancies
                WHERE (name ILIKE '%data scientist%' 
                or name ILIKE '%data science%' 
                or name ILIKE '%исследователь данных%' 
                or name ILIKE '%ml%' and name NOT LIKE '%HTML%' and name not like '%QML%' 
                or name ILIKE '%machine learning%' 
                or name ILIKE '%машинн%обучен%') 
                and(key_skills ILIKE '%python%')
            '''

In [44]:
df_19 = pd.read_sql_query(query_6_4, connection)
df_19

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2232680158.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_19 = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
query_6_5 = f'''SELECT
                round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1),2)
                FROM vacancies
                WHERE ((name ILIKE '%data scientist%')
                or (name ILIKE '%data science%')
                or (name ILIKE '%исследователь данных%')
                or (name LIKE '%ML%' and (name NOT LIKE '%HTML%'))
                or (name ILIKE '%machine learning%')
                or (name ILIKE '%машинн%обучен%')
                and (key_skills is not NULL))
            '''


In [46]:
df_20 = pd.read_sql_query(query_6_5, connection)
df_20

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1109550251.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_20 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
query_6_6 = f'''SELECT
                experience,
                ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) avg_salary
                FROM vacancies
                WHERE ((name ILIKE '%data scientist%')
                or (name ILIKE '%data science%')
                or (name ILIKE '%исследователь данных%')
                or (name LIKE '%ML%' and (name NOT LIKE '%HTML%'))
                or (name ILIKE '%machine learning%')
                or (name ILIKE '%машинн%обучен%'))
                and ((salary_from is not Null) 
                or (salary_to is not NULL))
                group by 1
                order by 2
            '''

In [48]:
df_21 = pd.read_sql_query(query_6_6, connection)
df_21

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/3552479725.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_21 = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

**Выводы по предметному анализу:**
1. К данным имеют отношение 1771 вакансия.
2. Для начинающего дата-сайентиста подходит 51 вакансия. 
3. 229 вакансия с требуемым ключевым навыком SQL или Postgres.
4. 357 вакансий для DS с требованием знания Python.
5. Примерно 6,4 ключевых навыков в среднем указано в вакансиях для DS.
6. Для DS зарплата для начинающего специалиста (без опыта) в среднем составляет 75тр, для специалиста с опытом 1-3 года в среднем 140тр, для специалиста с опытом 3-6 лет составляет 243тр.

**Дополнительные исследования:**

Составим таблицу, показывающую рейтинг регионов по средней зарплате для DS (беря в расчет регионы где вакансий больше чем одна) с обозначением наиболее популярного требования по опыту работы.

In [49]:
query_7_1 = f'''select
                a.name,
                count(v.id),
                ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) avg_salary,
                max(v.experience) most_popular_experience
                from vacancies v
                left join areas a ON v.area_id = a.id
                WHERE ((v.name ILIKE '%data scientist%')
                or (v.name ILIKE '%data science%')
                or (v.name ILIKE '%исследователь данных%')
                or (v.name LIKE '%ML%' and (v.name NOT LIKE '%HTML%'))
                or (v.name ILIKE '%machine learning%')
                or (v.name ILIKE '%машинн%обучен%'))
                and ((salary_from is not Null) 
                or (salary_to is not NULL))
                group by 1
                having count(v.id) > 1
                order by 3 desc
            '''

In [50]:
df_22 = pd.read_sql_query(query_7_1, connection)
df_22

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/1549919614.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_22 = pd.read_sql_query(query_7_1, connection)


name  count  avg_salary most_popular_experience
0          Армения      3    268863.0           От 3 до 6 лет
1           Москва     27    207630.0           От 3 до 6 лет
2  Санкт-Петербург      7    173335.0           От 3 до 6 лет
3      Новосибирск      3    166333.0           От 3 до 6 лет
4           Рязань      3    136667.0           От 3 до 6 лет
5     Екатеринбург      3    124000.0      От 1 года до 3 лет
6           Алматы      2     94587.0           От 3 до 6 лет
7      Владивосток      2     85000.0      От 1 года до 3 лет
8            Пермь      2     75000.0      От 1 года до 3 лет

Составим таблицу, показывающую распредение работодателей в зависимости от количества вакансий для DS без опыта работы, с обозначением средней зарплаты и регионов, где таких специалистов требуется больше всего.

In [51]:
query_7_2 = f'''select
                i.name,
                count(v.id)
                from industries i
                left join employers_industries ei on i.id = ei.industry_id
                left join employers e on ei.employer_id = e.id
                left join vacancies v on e.id = v.employer_id
                and ((salary_from is not Null) 
                or (salary_to is not NULL))
                group by 1
                order by 2 desc
                limit 15
            '''

In [52]:
df_23 = pd.read_sql_query(query_7_2, connection)
df_23

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/3522768962.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_23 = pd.read_sql_query(query_7_2, connection)


name  count
0                 Разработка программного обеспечения   4572
1   Системная интеграция,  автоматизации технологи...   4142
2   Интернет-компания (поисковики, платежные систе...   3387
3                                     Мобильная связь    742
4                                  Кадровые агентства    711
5                                  Интернет-провайдер    709
6                                Оптоволоконная связь    663
7                               Консалтинговые услуги    654
8                                 Фиксированная связь    589
9   Маркетинговые, рекламные, BTL, дизайнерские, E...    538
10                       Розничная сеть (продуктовая)    515
11                                               Банк    394
12                        Государственные организации    362
13                                   Интернет-магазин    326
14  Электронно-вычислительная, оптическая, контрол...    302

In [53]:
query_7_3 = f'''select
                e.name,
                count(v.id),
                ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) avg_salary,
                max(a.name)
                from employers e
                left join vacancies v on e.id = v.employer_id
                left join areas a on v.area_id = a.id
                WHERE ((v.name ILIKE '%data scientist%')
                or (v.name ILIKE '%data science%')
                or (v.name ILIKE '%исследователь данных%')
                or (v.name LIKE '%ML%' and (v.name NOT LIKE '%HTML%'))
                or (v.name ILIKE '%machine learning%')
                or (v.name ILIKE '%машинн%обучен%'))
                and ((salary_from is not Null) 
                or (salary_to is not NULL))
                and v.experience = 'Нет опыта'
                group by 1
                order by 2 desc
                limit 15
            '''

In [54]:
df_24 = pd.read_sql_query(query_7_3, connection)
df_24

/var/folders/m1/4hd9v_6d76sgg850qhp855t40000gn/T/ipykernel_2237/2667247961.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_24 = pd.read_sql_query(query_7_3, connection)


name  count  avg_salary  \
0                                  DNS Головной офис      1     70000.0   
1                                            Апэрбот      1     80000.0   
2  КБ Уральский банк реконструкции и развития (УБ...      1     37000.0   
3                                           ННФормат      1     98000.0   
4                                  Строительный Двор      1     72500.0   
5                                      Улыбка радуги      1     50000.0   
6                                   Университет ИТМО      1    115000.0   

               max  
0      Владивосток  
1      Новосибирск  
2     Екатеринбург  
3  Санкт-Петербург  
4           Москва  
5  Санкт-Петербург  
6  Санкт-Петербург

# Общий вывод по проекту

***Наибольшее число специалистов требуется в городах с населением выще 1млн человек, работающих полный день и примерно половина вакансий для специаслистов без опыта работы.
Компания Яндекс занимает первое место по числу вакансий в каждом городе-миллионнике. IT отрасль, а именно разработска программного обеспечения, занимает первное место по числу вакансий из предствленных.
Из отрасли IT значительное место занимают вакансии имеющие отношения к данным, большее число из которых подходит для имеющих минимальный (и выше) опыт работы.
Минимальный опыт работы чаще встречается в вакансиях из не столичных и не иностранных регионов.
В целом представленный нобор данных говорит о востребованности специалистов в IT сфере больше чем в других.***
